### PR TITLE
Milestone 1

### PR DESCRIPTION
**Pull request by sbchaos**

### CODE REVIEW

File: src/main/java/com/gojek/parkinglotproject/Car.java [hunk 1]
```diff
@@ -0,0 +1,21 @@
+package com.gojek.parkinglotproject;
+
+public class Car {
```
Comment (Waseem1509) [inline src/main/java/com/gojek/parkinglotproject/Car.java:3]: use lombok
@Getter

File: src/main/java/com/gojek/parkinglotproject/Ticket.java [hunk 2]
```diff
@@ -0,0 +1,36 @@
+package com.gojek.parkinglotproject;
+
+public class Ticket {
+    private final int number;
+    private Car car;
```
Comment (Waseem1509) [inline src/main/java/com/gojek/parkinglotproject/Ticket.java:5]: Use only Car DTO as an attribute of ticket here.

File: src/main/java/com/gojek/parkinglotproject/Ticket.java [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+    private String carColor;
+    private String carRegistrationNumber;
```
Comment (Waseem1509) [inline src/main/java/com/gojek/parkinglotproject/Ticket.java:7]: Should not create another variable with same names and meaning to access car's attributes inside ticket

File: src/main/java/com/gojek/parkinglotproject/Ticket.java [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+
+    public Ticket(int number, Car car){
```
Comment (Waseem1509) [inline src/main/java/com/gojek/parkinglotproject/Ticket.java:9]: Indentation of bracket

File: src/main/java/com/gojek/parkinglotproject/Ticket.java [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+    public void parkCar(Car car) {
+        this.car = car;
+    }
+    public boolean isEmpty() {
+        return car == null;
+    }
+    public int getTicketNumber() {
+        return number;
+    }
+    public String getCarColor() {
```
Comment (Waseem1509) [inline src/main/java/com/gojek/parkinglotproject/Ticket.java:25]: we dont need this getter

File: src/main/java/com/gojek/parkinglotproject/Ticket.java [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+        return this.carColor;
+    }
+    public String getCarRegistrationNumber() {
```
Comment (Waseem1509) [inline src/main/java/com/gojek/parkinglotproject/Ticket.java:28]: dont need this getter

File: src/main/java/com/gojek/parkinglotproject/Ticket.java [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+        return this.carRegistrationNumber;
+    }
+
+    public void removeCar() {
```
Comment (Waseem1509) [inline src/main/java/com/gojek/parkinglotproject/Ticket.java:32]: This is absurd as we cannot assign same ticket to another car

File: src/main/java/com/gojek/parkinglotproject/ParkingLot.java [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+    }
+
+    public void park(String registrationNumber, String color) {
+        Car car = new Car(registrationNumber, color);
+        if (issuedTicketsCount == size) {
+            System.out.println("Sorry, parking lot is full");
+            return;
+        }
+        int num;
+        for (num = 1; num <= size; num++) {
```
Comment (Waseem1509) [inline src/main/java/com/gojek/parkinglotproject/ParkingLot.java:58]: this should be for (int num = 0; num < size; num++)

Comment (Waseem1509) [inline src/main/java/com/gojek/parkinglotproject/ParkingLot.java:58]: NIT PICK initialize num in the loop itself for (int num ...

File: src/main/java/com/gojek/parkinglotproject/Ticket.java [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+public class Ticket {
+    private final int number;
+    private Car car;
+    private String carColor;
+    private String carRegistrationNumber;
+
+    public Ticket(int number, Car car){
+        this.number = number;
+        this.car = car;
+        this.carColor = car.getCarColor();
```
Comment (Waseem1509) [inline src/main/java/com/gojek/parkinglotproject/Ticket.java:12]: dont need these assignments .. redundant.

File: src/main/java/com/gojek/parkinglotproject/Ticket.java [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+        this.carRegistrationNumber = car.getCarRegistrationNumber();
+    }
+
+    public void parkCar(Car car) {
+        this.car = car;
+    }
+    public boolean isEmpty() {
```
Comment (Waseem1509) [inline src/main/java/com/gojek/parkinglotproject/Ticket.java:19]: add a function to return this object's car object value

File: src/main/java/com/gojek/parkinglotproject/ParkingLot.java [hunk 11]
```diff
@@ -0,0 +1,130 @@
...
+        return parkingNumber.get(num).isEmpty();
+    }
+
+
+    public void leave(int ticketNumber) {
+        if (issuedTicketsCount == 0)
+            System.out.println("Parking lot is already empty");
+        else if (!validateParkingNumber(ticketNumber))
+            System.out.println("Enter a valid ticket number between 1 and " + size);
+        else if (checkEmpty(ticketNumber))
```
Comment (Waseem1509) [inline src/main/java/com/gojek/parkinglotproject/ParkingLot.java:41]: we can keep this check inside validateParkingNumber itself.

File: src/main/java/com/gojek/parkinglotproject/ParkingLot.java [hunk 12]
```diff
@@ -0,0 +1,130 @@
...
+
+    public int getSize() {
+        return size;
+    }
+
+    public boolean validateParkingNumber(int num) {
+        return num >= 1 && num <= size;
+    }
+
+    public boolean checkEmpty(int num) {
```
Comment (Waseem1509) [inline src/main/java/com/gojek/parkinglotproject/ParkingLot.java:31]: dont need separate function for checking empty

Comment (Waseem1509) [inline src/main/java/com/gojek/parkinglotproject/ParkingLot.java:31]: we can keep this in above function itself.

File: src/main/java/com/gojek/parkinglotproject/ParkingLot.java [hunk 13]
```diff
@@ -0,0 +1,130 @@
...
+        return parkingNumber.get(num).isEmpty();
```
Comment (Waseem1509) [inline src/main/java/com/gojek/parkinglotproject/ParkingLot.java:32]: should be num - 1

File: src/main/java/com/gojek/parkinglotproject/ParkingLot.java [hunk 14]
```diff
@@ -0,0 +1,130 @@
...
+
+    public void park(String registrationNumber, String color) {
+        Car car = new Car(registrationNumber, color);
+        if (issuedTicketsCount == size) {
+            System.out.println("Sorry, parking lot is full");
+            return;
+        }
+        int num;
+        for (num = 1; num <= size; num++) {
+            if (parkingNumber.get(num).isEmpty()) {
```
Comment (Waseem1509) [inline src/main/java/com/gojek/parkinglotproject/ParkingLot.java:59]: num - 1 should be there

File: src/main/java/com/gojek/parkinglotproject/ParkingLot.java [hunk 15]
```diff
@@ -0,0 +1,130 @@
...
+                parkingNumber.get(num).parkCar(car);
```
Comment (Waseem1509) [inline src/main/java/com/gojek/parkinglotproject/ParkingLot.java:60]: num - 1 should be there

File: src/main/java/com/gojek/parkinglotproject/ParkingLot.java [hunk 16]
```diff
@@ -0,0 +1,130 @@
...
+    }
+
+    public void status() {
+        if (issuedTicketsCount == 0)
+            System.out.println("Parking lot is empty");
+        else {
+            String status = "#Ticket No. #Registration No. #Colour\n";
+            for (int num = 1; num <= size; num++) {
+                if (!parkingNumber.get(num).isEmpty())
+                    status += num + "\t" + parkingNumber.get(num).getCarRegistrationNumber() + "\t" + parkingNumber.get(num).getCarColor() + "\n";
```
Comment (Waseem1509) [inline src/main/java/com/gojek/parkinglotproject/ParkingLot.java:75]: num - 1 should be there

File: src/main/java/com/gojek/parkinglotproject/ParkingLot.java [hunk 17]
```diff
@@ -0,0 +1,130 @@
...
+            }
+            status.substring(0, status.length() - 1);
```
Comment (Waseem1509) [inline src/main/java/com/gojek/parkinglotproject/ParkingLot.java:77]: not sure why this is being used.

File: src/main/java/com/gojek/parkinglotproject/ParkingLot.java [hunk 18]
```diff
@@ -0,0 +1,130 @@
...
+            System.out.println(status);
+        }
+    }
+
+    public void registrationNumbersForCarsWithColour(String color) {
+        String regNumsOfCarsWithColor = "";
+        for (Ticket num : parkingNumber) {
+            if (!num.isEmpty() && num.getCarColor().equals(color)) {
```
Comment (Waseem1509) [inline src/main/java/com/gojek/parkinglotproject/ParkingLot.java:85]: should be num.getCar().getCarColor

File: src/main/java/com/gojek/parkinglotproject/ParkingLot.java (hunk 16 above)
Comment (Waseem1509) [inline src/main/java/com/gojek/parkinglotproject/ParkingLot.java:75]: should be parkingNumber.get(num).getCar().getCarRegistrationNumber()

File: src/main/java/com/gojek/parkinglotproject/ParkingLot.java (hunk 17 above)
Comment (Waseem1509) [reply src/main/java/com/gojek/parkinglotproject/ParkingLot.java:77]: I think its correct to remove extra new line char

File: src/main/java/com/gojek/parkinglotproject/ParkingLot.java [hunk 19]
```diff
@@ -0,0 +1,130 @@
...
+        if (issuedTicketsCount == 0)
+            System.out.println("Parking lot is empty");
+        else {
+            String status = "#Ticket No. #Registration No. #Colour\n";
+            for (int num = 1; num <= size; num++) {
+                if (!parkingNumber.get(num).isEmpty())
+                    status += num + "\t" + parkingNumber.get(num).getCarRegistrationNumber() + "\t" + parkingNumber.get(num).getCarColor() + "\n";
+            }
+            status.substring(0, status.length() - 1);
+            System.out.println(status);
```
Comment (Waseem1509) [inline src/main/java/com/gojek/parkinglotproject/ParkingLot.java:78]: we can print status.substring(0, status.length() - 1) directly here.

File: src/main/java/com/gojek/parkinglotproject/ParkingLot.java [hunk 20]
```diff
@@ -0,0 +1,130 @@
...
+        String regNumsOfCarsWithColor = "";
+        for (Ticket num : parkingNumber) {
+            if (!num.isEmpty() && num.getCarColor().equals(color)) {
+                regNumsOfCarsWithColor += num.getCarRegistrationNumber() + ", ";
+            }
+        }
+        if (regNumsOfCarsWithColor.equals(""))
+            System.out.println("Not found");
+        else {
+
```
Comment (Waseem1509) [inline src/main/java/com/gojek/parkinglotproject/ParkingLot.java:92]: unnecessary new line in code should be avoided

File: src/main/java/com/gojek/parkinglotproject/ParkingLot.java [hunk 21]
```diff
@@ -0,0 +1,130 @@
...
+            System.out.println(regNumsOfCarsWithColor.substring(0, regNumsOfCarsWithColor.length() - 2));
+        }
+
+    }
+
+    public void slotNumbersForCarsWithColour(String color) {
+        String slotOfCarsWithColor = "";
+        for (Ticket num : parkingNumber) {
+            if (!num.isEmpty() && num.getCarColor().equals(color)) {
```
Comment (Waseem1509) [inline src/main/java/com/gojek/parkinglotproject/ParkingLot.java:101]: should be num.getCar.getCarColor()

File: src/main/java/com/gojek/parkinglotproject/ParkingLot.java [hunk 22]
```diff
@@ -0,0 +1,130 @@
...
+                slotOfCarsWithColor += num.getTicketNumber() + ", ";
+            }
+        }
+        if (slotOfCarsWithColor.equals(""))
+            System.out.println("Not found");
+        else {
```
Comment (Waseem1509) [inline src/main/java/com/gojek/parkinglotproject/ParkingLot.java:107]: no need braces here

File: src/main/java/com/gojek/parkinglotproject/ParkingLot.java [hunk 23]
```diff
@@ -0,0 +1,130 @@
...
+        }
+
+    }
+
+    public void slotNumberForRegistrationNumber(String registrationNumber) {
+        String slotNum = "";
+
+
+        for (int num = 1; num <= size; num++) {
+            if (!parkingNumber.get(num).isEmpty() && parkingNumber.get(num).getCarRegistrationNumber().equals(registrationNumber)) {
```
Comment (Waseem1509) [inline src/main/java/com/gojek/parkinglotproject/ParkingLot.java:118]: should be num - 1

Comment (Waseem1509) [reply src/main/java/com/gojek/parkinglotproject/ParkingLot.java:118]: and should be parkingNumber.get(num - 1).getCar().getCarRegistrationNumber()

File: src/main/java/com/gojek/parkinglotproject/Main.java [hunk 24]
```diff
@@ -0,0 +1,95 @@
...
+        return false;
+    }
+
+    public void startInteractiveSession() {
+        String command = "";
+        while (!command.equals("quit")) {
+            System.out.print("Input your command : ");
+            Scanner scanner = new Scanner(System.in);
+            String tempInput = scanner.nextLine();
+            tempInput.replace("\n", "");
```
Comment (Waseem1509) [inline src/main/java/com/gojek/parkinglotproject/Main.java:25]: dont need to replace new line as there will be none

File: src/main/java/com/gojek/parkinglotproject/Main.java [hunk 25]
```diff
@@ -0,0 +1,95 @@
...
+            String[] inputs = scanner.nextLine().split(" ");
+            String command = inputs[0];
+            commandOutputs(inputs);
+        }
+    }
+
+
+    public void commandOutputs(String[] inputs) {
+        String command = inputs[0];
+        if (command.equals(commandsList[0]) && inputs.length == 2)
```
Comment (Waseem1509) [inline src/main/java/com/gojek/parkinglotproject/Main.java:53]: This is not very scalable and not easily extensible for any new commands coming in. We can use strategy design pattern to implement multiple forms of operations on the parking lot

File: src/test/java/com/gojek/parkinglotproject/TicketTest.java [hunk 26]
```diff
@@ -0,0 +1,42 @@
+package com.gojek.parkinglotproject;
+import org.junit.Test;
+import static org.junit.jupiter.api.Assertions.*;
+
+public class TicketTest {
+
+    private final String registrationNumber = "KA-01-HH-1234";
+    private final String color = "Black";
+    private final Car car = new Car(registrationNumber, color);
```
Comment (Waseem1509) [inline src/test/java/com/gojek/parkinglotproject/TicketTest.java:9]: directly assigning some test values in the constructor

File: src/test/java/com/gojek/parkinglotproject/TicketTest.java [hunk 27]
```diff
@@ -0,0 +1,42 @@
...
+
+    @Test
+    public void testInitializationTicket() {
+        Ticket ticket = new Ticket(1, car);
+        assertEquals(registrationNumber, ticket.getCarRegistrationNumber());
```
Comment (Waseem1509) [inline src/test/java/com/gojek/parkinglotproject/TicketTest.java:14]: should be ticket.getCar().getCarRegistrationNumber()

**Comment by sbchaos** [comment]
Done

**Pull request closed without merging**